### ISSUE TITLE
How to implement pad array based on boolean mask?

### ISSUE THREAD

**Issue by thisiscam** [google/jax#1625]
I was wondering how to implement the following piece of logic, probably found often in LSTM/RNN, though my use case is not one of those:

I have two input arrays:
```
groups = np.array([0, 1, 2, 2, 3, 2, 1]) # integer array, each element in range K (say K = 4)
X = jax.random.normal(key, groups.shape) # feature array, same length as group
```

Now I want to group `X` based on `groups`, where same integer in `groups` means the corresponding indices of `X` belongs to the same group. Apparently, groups will be of different length and AFAIK Jax/XLA cannot handle dynamic sizes natively (unless we compile for many times for each size); however, what I actually need is to group X, and fill into a zero-padded array, where each size of group has being padded to some upper bound that is statically fixed. 
 
In Numpy one way to specify this is:
```
X_groups = []
for k in range(K):
  X_group = X[group_labels == k]   # line a)
  X_group = np.broadcast_to(X, (GROUP_SIZE_UPPER_BOUND, )).  # line b)
  X_groups.append(X_group)
X_groups = np.stack(X_groups)
```

I was wondering if there's a good way to implement this logic in Jax, pushing as much (jitt'ed) computation to GPU as possible? 

The main issue with the above Numpy code is that, line a) has dynamic boolean masking which is not supported by JIT; however line b) immediately turns the array back to static shape. I'm hoping  someone can point me out that the combination of the two is actually (magically) possible in Jax :)

**Comment by levskaya** [comment]
I think you're wanting to implement a "group-by" operation and it _is_ a little trickier to do in a system with static shape constraints.   
One way to do it: there are a few loop constructs in JAX that can be used to easily build the desired output, here's an implementation with a lax.scan loop (a fori_loop could make the below a bit simpler at the cost of an explicit loop index.  See those functions' docstrings for their api.):
```python
K = 4
GROUP_SIZE_UPPER_BOUND = 4
key = jax.random.PRNGKey(0)

# inputs
groups = np.array([0, 1, 2, 2, 3, 2, 1])
X = jax.random.normal(key, groups.shape)

# initialize outputs
X_grouped = onp.zeros((K, GROUP_SIZE_UPPER_BOUND))
group_cnts = np.zeros((K,), np.int32)

def group_by(data, var):
  x, g = var
  x_grouped, group_cnts = data
  # append entries into specified group
  x_grouped = jax.ops.index_add(x_grouped, (g, group_cnts[g]), x)
  # track how many entries appended into each group
  group_cnts = jax.ops.index_add(group_cnts, g, 1)
  return (x_grouped, group_cnts), 0 # '0' is just a dummy value

# run scan, result is stored in the carry rather than accumulated scan output
# could simplify with a fori_loop instead:
(X_grouped, group_cnts), _ = lax.scan(group_by, 
                                     (X_grouped, group_cnts), # initial state
                                     (X, groups)) # data to loop over

X_groups
```

this example in a colab:
https://colab.research.google.com/drive/15auuAAK7UAgIry42kSCRo5y3EPeFr0mB

**Comment by thisiscam** [comment]
Amazing! Thanks! I will adapt this code and post my final solution.

**Comment by hawkinsp** [comment]
Closing due to lack of activity. Feel free to reopen if there's more to discuss!
